### ISSUE TITLE
PATH not updated correctly for portable package install

### ISSUE THREAD

**Issue by jamiebarrow** [microsoft/winget-cli#2498]
### Brief description of your issue

When installing the package `stedolan.jq` which is a `portable` package type, winget appears to add the `%LOCALAPPDATA%\Microsoft\WinGet\Links` directory to the user's `PATH` environment variable, however, running a new user command prompt instance does not reflect this updated `PATH` value.

As the location is not on the path, it is then not possible to run the installed executable from the new command prompt instance.

The `PATH` does seem to be updated correctly for new elevated user command prompts, but not normal user command prompts.

### Steps to reproduce

1. Ensure that the `%APPLOCALDATA%\Microsoft\WinGet\Links` directory is not already on the PATH (both for the user, and the machine)
    - **Note:** I recommend doing this via the `Control Panel` > `System` (or `Settings` > `System` > `About`) > `Advanced system settings` > `Environment Variables...`, to ensure the environment variables are actually applied
    - `echo %PATH% | findstr /I %LOCALAPPDATA%\Microsoft\WinGet\Links` should not return results from a normal user command prompt
2. In an elevated command prompt or Powershell, run `winget install --id stedolan.jq`
    - it should be `Successfully installed` and also give a message regarding `Path environment variable modified; restart your shell to use the new value.`
    - `echo %PATH% | findstr /I %LOCALAPPDATA%\Microsoft\Winget\Links` should not return results from this elevated command prompt

### Expected behavior

3. Open an elevated command prompt, and confirm the `PATH` has been updated for elevated users
    - `echo %PATH% | findstr /I %LOCALAPPDATA%\Microsoft\Winget\Links` should return results from this elevated command prompt
    - `where jq` should locate `jq.exe` on the path in `%LOCALAPPDATA%\Microsoft\WinGet\Links\jq.exe` (which is a symlink to the actual exe, e.g. `%LOCALAPPDATA%\Microsoft\WinGet\Packages\stedolan.jq_Microsoft.Winget.Source_8wekyb3d8bbwe\jq-win64.exe`)
    - running `jq` should be possible as it is on the path
4. Open a normal user command prompt, and confirm the value of the PATH environment variable should have `%LOCALAPPDATA%\Microsoft\WinGet\Links`
    - `echo %PATH% | findstr /I %LOCALAPPDATA%\Microsoft\Winget\Links` should return result
    - `where jq` should find `jq.exe` on the path
    - running the `jq` command on the path should be possible

### Actual behavior

4. Open a normal user command prompt, and confirm the value of the PATH environment variable should have `%LOCALAPPDATA%\Microsoft\WinGet\Links`, **but does not**
    - `echo %PATH% | findstr /I %LOCALAPPDATA%\Microsoft\Winget\Links` should return result, but is empty
    - `where jq` does not find `jq.exe` on the path
    - running the `jq` command on the path is not possible

5. Go to `Control Panel` > `System` (or `Settings` > `System` > `About`) > `Advanced system settings` > `Environment Variables...`
6. Under `User variables for %USERNAME%`, edit the value for `Path`, and note that indeed the `%LOCALAPPDATA%\Microsoft\WinGet\Links` entry *has already been added* to the end of the `Path` variable
7. Press `OK` to close the `Edit environment variable` dialog, and `OK` again to close the `Environment Variables` dialog, which should then apply these environment variables to new command prompts
8. Launch a new normal user command prompt, and confirm the value of the `PATH` environment variable has **only now been updated** to include `%LOCALAPPDATA%\Microsoft\WinGet\Links`
    - `echo %PATH% | findstr /I %LOCALAPPDATA%\Microsoft\Winget\Links` returns expected result
    - `where jq` finds `jq.exe` on the path
    - running the `jq` command is only now possible

I believe I also tried with `winget install --scope user` and had the same results.

The experience was also the same in both `Command Prompt` and in `Windows PowerShell` v5.1, and also same experience when run via `Windows Terminal` versus plain old `Windows Console Host` (i.e. `Settings` > `Privacy & security` > `For developers` > `Terminal`).

### Environment

I managed to reproduce this on a Windows 10 machine, as well as Windows 11 machine. Below is the winget info for the Windows 11 machine:

```shell
Windows Package Manager v1.3.2091
Copyright (c) Microsoft Corporation. All rights reserved.

Windows: Windows.Desktop v10.0.22000.856
System Architecture: X64
Package: Microsoft.DesktopAppInstaller v1.18.2091.0
```

PowerShell details:

```shell
PS C:\Users\james> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      5.1.22000.832
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.22000.832
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

**Comment by ryfu-msft** [comment]
Hi @jamiebarrow,

This is actually by design. We are unable to create symlinks when a user does not have admin rights or does not have developer mode enabled. We instead will add the install directory directly to PATH so that the executable can still be run from the command prompt. If you did not specify a location during install, the default install location should be `%LOCALAPPDATA%\Microsoft\Winget\Packages\Package_Source`. Can you verify that this was added to PATH instead and that you are still able to run the exe from the command prompt.

**Comment by jamiebarrow** [comment]
> We are unable to create symlinks when a user does not have admin rights or does not have developer mode enabled

I was running as Administrator when running `winget install --id stedolan.jq`. The symlink `%LOCALAPPDATA%\Microsoft\WinGet\Links\jq.exe` is created successfully, as explained initially. I then closed the Administrator command prompt.

Running a completely new _user command prompt_, which was then not running as Administrator, should have had the updated `PATH` applied. Inspecting the value of `%PATH%` in the new user command prompt did not show `%LOCALAPPDATA%\Microsoft\WinGet\Links` on the `PATH`, and `where jq` also returned no results (`jq` was not on the path). However, when going to the Advanced System Settings dialog to see if the `PATH` environment variable had been changed (as the winget output did mention it had), it indeed did display the entry for the `%LOCALAPPDATA%\Microsoft\WinGet\Links` folder in the User portion of the dialog. This was baffling to me, since I've never seen that the user's environment variables are added, but not applied to new user shells.

Only after then pressing OK on the Advanced System Settings dialog(s), and then launching a new user command prompt, did `jq` successfully appear on the path. To be clear, the item was already there, and no modifications were done while in the Advanced System Settings dialog was open, purely used it to view the current value of `PATH`.

At no point was `%LOCALAPPDATA%\Microsoft\WinGet\Packages\stedolan.jq_Microsoft.Winget.Source_8wekyb3d8bbwe` (or similar) on the `PATH` either.

I managed to reproduce this behavior multiple times last night while creating the issues.

However, also strangely, I tried to reproduce it again tonight on my personal computer but it seemingly works as expected (also verified the `winget info` result is the same, so I believe using the same version of winget on the same build of Windows). I am not sure if there was any other kind of Windows update that could have explained this change in my experience.

I will try to reproduce it again on my work computer tomorrow, to see if it is still an issue there.

**Comment by jamiebarrow** [comment]
Just reproduced it on my work computer.

Steps are the same as I initially reported, but will re-list them here in case of confusion:
1. Launch Administrator PowerShell
2. `winget install --id stedolan.jq`
3. Close Administrator PowerShell
4. Open new user Command Prompt
5. `jq` is not on path, close this Command Prompt
6. Open Advanced System Settings > Environment Variables, Press 'OK' twice to close both dialogs, making no changes to `PATH` as it was already updated in step (2) above
7. Open new user Command Prompt
8. `jq` does execute correctly, as only now PATH is effective

Also reproduced the above when using `cmd` in step (1).

```
Windows Package Manager v1.3.2091
Copyright (c) Microsoft Corporation. All rights reserved.

Windows: Windows.Desktop v10.0.19042.1889
System Architecture: X64
Package: Microsoft.DesktopAppInstaller v1.18.2091.0
```

**Comment by jamiebarrow** [comment]
For reference, in the Administrator PowerShell, after running `winget install`, I verified that the `PATH` user environment variable was updated with the `Links` directory:

```pwsh
[System.Environment]::GetEnvironmentVariable("PATH", "User")
```

But then when closing the Admin PowerShell, and opening a new user command prompt/shell (`cmd` or `pwsh`) it is not in effect, as if it was updated but has not been applied to the user's "session", whereas I usually would expect if a user environment variable is updated then all new command prompt instances should get these updated values.

To me, it would make sense that it would get applied to the user's current "session", especially when in the Administrator PowerShell it mentions `Path environment variable modified; restart your shell to use the new value.`, this to me would intuitively mean I just need to launch a new command prompt, but is not the case.

To me it would be even better if the same kind of experience you get with the Chocolatey `refreshenv` - [see here](https://docs.chocolatey.org/en-us/create/functions/update-sessionenvironment)

**Comment by yilmazdurmaz** [comment]
Although the title might be misleading, I can confirm the situation even for machine-scoped installation. it is more like "PATH update is not triggered for users"

in the linked issue above, I stumbled upon this "PATH not updated" issue. I thought it was a PATH installation issue but it turned out that the new updated PATH was not used for the user.
- upon installing a package (gnu.nano) with **machine scope**, `c:\Program Files\WinGet\Links\` is added to system PATH
- user cannot see this change in the path and thus cannot use the installed package

This might be an issue on the first run as I believe the path would update itself on a system reset but that should not be an excuse for trying to find what causes this update failure. (PS: It might even be related to a recent system update, so heads up)

**Comment by stephengillie** [comment]
As a general behavior, Command Prompt and PowerShell windows appear to only update their respective Path variables at launch. I've been seeing this when Git.Git is a dependency, because it's also found through the Windows Path. 
* https://github.com/microsoft/winget-pkgs/issues/90650

**Comment by ezgambac** [comment]
This issue made me lost an hour of work until I found this thread.
I can confirm that installing jq, helm, kubelogin won't work unless you install from admin AND reload the path manually, multiple new windows terminals/powershell sessions didn't work.
I've added the following command to my pwsh profile to work around this issue, but a package manager can't screw this up:

$env:Path = [System.Environment]::GetEnvironmentVariable("Path","Machine") + ";" + [System.Environment]::GetEnvironmentVariable("Path","User")

**Comment by danielfdickinson** [comment]
> Although the title might be misleading, I can confirm the situation even for machine-scoped installation. it is more like "PATH update is not triggered for users"
> 
> in the linked issue above, I stumbled upon this "PATH not updated" issue. I thought it was a PATH installation issue but it turned out that the new updated PATH was not used for the user.
> 
> * upon installing a package (gnu.nano) with **machine scope**, `c:\Program Files\WinGet\Links\` is added to system PATH
> * user cannot see this change in the path and thus cannot use the installed package
> 
> This might be an issue on the first run as I believe the path would update itself on a system reset but that should not be an excuse for trying to find what causes this update failure. (PS: It might even be related to a recent system update, so heads up)

I believe this is being tracked in a separate issue - #2993

**Comment by ryfu-msft** [comment]
This issue has been resolved by #3002. The fix now adds the install directory to PATH instead of the link directory if a symlink cannot be created (user mode/ developer mode off).

I tried this out in the latest stable client version (1.6.2771) with developer mode disabled and no prior portable installations. I was able to successfully run the command `jq` after restarting my shell.

**Comment by yilmazdurmaz** [comment]
Hi @ryfu-msft 

main post is too long and I forgot what it was (lazy mode on). However, I have one thing to add:

(1.6.2771) upon installing GNU.Nano, "Windows Terminal" sees it without any further action, but "cmd" or "powershell" consoles don't. for them to work, I open the "system properties -> environment variables" page, then just close it without any edit so they too see the installation.
